### PR TITLE
feat: add document retries

### PR DESCRIPTION
**Pull request by kruskall**
~Blocked by https://github.com/elastic/go-docappender/pull/100 (to make testing easier)~

~Add document retries logic. Followup to https://github.com/elastic/go-docappender/pull/91 (important to understand how the compression logic changed)~

~### Compression off:~

~In this case we simply iterate over the array and submit the failed events to the buffer. We can count them using newlines.~

~### Compression on:~

~Retrieve the offset of the failed event and decompress the "gzip container". Loop over the array and submit the failed event to buffer.~

### Testing

This PR adds tests to cover as mush edge cases as possible: one event for each gzip container, multiple events per gzip container, all events in a gzip container. Tests also verify the retry logic twice to ensure there are no conflicts between flushes.

Update:

### Changes:

the approach was simplified: iterate over the array and resubmit the failed events to the buffer. For gzip we read in batches to avoid decompress the whole request.

### CODE REVIEW

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:273]: Could we reuse this buffer after its contents are written to the `gzip.Writer`? We're creating at least a 1KB `[]byte` for each failed item. If that buffer is expanded, I imagine it will be more efficient to recycle the byte slice after writing to the gzip compressor.

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:274]: Wouldn't it suffice to `gr.Read(buf)`?

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:293]: The current test cases don't seem to cover the contents of the two `for` clauses. Would you mind adding a test or two where the bulk item is large enough to require more capacity?

**Review by marclop — COMMENTED** [review]
I've noticed that the tests have `MaxRequests: 1`, which is why the tests are passing. However, in a more realistic scenario with `MaxRequests > 1`, the documents stored in the local `bulk_indexer` won't be flushed to Elasticsearch until the bulk indexer is cycled through the channel.

https://github.com/elastic/go-docappender/blob/c8c4e54dbd8820190e465c0811122cb6ffbde24d/appender.go#L470-L479

We should prevent that from happening. I think we may need to introduce a new channel for bulk_indexers that have items from previous 429 failures, and treat it with higher priority than `a.available`. There are some edge cases we need to handle (like closing the appender), but it would be better than simply sending the bulk_indexer with cached items back to the channel like empty indexers.

Another concern is that there's currently no limit on the number of times `429s` will be retried. In theory, the entire buffer could be filled with documents that have not been indexed due to `429s`, which would cause only 1 new event to be sent in each subsequent flush. Have you given any thought on how we could configure an upper limit of consecutive retries?

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:143]: I think it looks a bit more consistent and concise this way. WDYT?
```suggestion
func newBulkIndexer(client *elasticsearch.Client, compressionLevel int, maxDocRetry int) *bulkIndexer {
	b := &bulkIndexer{
		client:           client,
		maxDocumentRetry: maxDocRetry,
		retryCounts:      make(map[int]int),
	}
```

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:230]: Can it be negative?
```suggestion
	if b.maxDocumentRetry > 0 {
```

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:276]:
```suggestion
		// Eliminate previous retry counts that aren't present in the last bulk
		// request response.
```

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:281]: Do we need to keep iterating if the position matches?
```suggestion
					found = true
					break
```

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:348]: Can it be negative?
```suggestion
		if len(tmp) > 0 {
```

File: config.go
Comment (marclop) [inline config.go:59]: Should this default to 3?

File: bulk_indexer.go
Comment (marclop) [inline bulk_indexer.go:301]: I think a couple of comments would go a long way:
```go
				// Increment 429 count for the positions found.
				count := b.retryCounts[res.Position] + 1
				if count > b.maxDocumentRetry {
					tmp = append(tmp, res)
					continue
				}
				// Since some items may have succeeded, counter positions need
				// to be updated to match the next current buffer position.
				b.retryCounts[b.itemsAdded] = count
```

I like the retry tracking algorithm, it's neat!

**Review by marclop — COMMENTED** [review]
> To be honest, I think this is fine. We already changed quite a lot of things because of concerns around complexity. If ES is returning 429s there are a lot of events buffered/coming in. I don't think retried events are gonna stay "idle" for a long time.

I don't think it's acceptable to do all of this retry work and then return a potentially full `bulkIndexer` back to the `available` channel, where it may or may not be flushed in the future.

While it may be the case that 429s tend to happen on higher throughputs and the `bulkIndexer` may not remain in that channel for longer, it won't be flushed on `appender.Close()`, leading to those buffered events to be completely lost.

Could you please add a test that ensures that previously buffered 429s are flushed once more on `indexer.Close()`?

In the case where 429s are a considerable % of the total documents in a bulk request, say 30-40%, we'd leave a bunch of half-full (or almost entirely full if 429 % is even higher) `bulkIndexers` in that `available` channel.

> I think we should get rid of the "active" channel entirely and let the runtime handle the channels but that should go in a separate PR.

Could you elaborate on this? I'm not following your reasoning and how the runtime would be of any help to us in this case.

**Comment by simitt** [comment]
The `bench-diff` check is green, but looking at the [logs](https://github.com/elastic/go-docappender/actions/runs/7783017661/job/21220601839?pr=99), some benchmark errors happened; please take a look and either fix the issue or loop in the automation team, in case it's not legit errors.

File: appender.go
Comment (marclop) [inline appender.go:201]: Returning early will cause the non-empty `bulkIndexers` to not be flushed. What do you think about the snippet below?
```go
	// Wait until all active indexers have been flushed.
	err := a.errgroup.Wait()
	close(a.available)
	for bi := range a.available {
		a.errgroup.Go(func() error {
			if e := a.flush(context.Background(), bi); e != nil {
				return fmt.Errorf("failed to flush events on close: %w", e)
			}
			return nil
		})
	}
	return errors.Join(err, a.errgroup.Wait())
```

File: bulk_indexer.go
Comment (marclop) [reply bulk_indexer.go:373]: That was what my comment was about, rather than `!=`, it seems more appropriate to use the `>` operator.

**Review by marclop — COMMENTED** [review]
Changes mostly LGTM, I think we need to retry flushing even if the error group returns an error.

> > Could you elaborate on this? I'm not following your reasoning and how the runtime would be of any help to us in this case.
> 
> Ideally, we would just read from the channel and let the runtime switch between them, removing the `active` channel completely. The flush interval would then cause a flush request for the bulk indexers that still have events in the buffer. The issue of bulk indexers remaining in a "limbo" would not apply anymore because they would all be active.

I still don't follow what you meant.

Also, have you run some benchmarks using the `tilt` environment and monitoring how the metrics look? I think it's worth doing so before merging.

File: bulk_indexer.go
Comment (simitt) [reply bulk_indexer.go:373]: +1 on using `>` for better readability when only dealing with positive numbers.

File: appender.go
Comment (simitt) [reply appender.go:201]: Could you open a follow up issue then to tackle this? With adding retry behavior we are moving more towards reliability vs. dropping events, so the ask to change the behavior for flushing non empty bulk indexers seems totally fair.

**Comment by simitt** [comment]
>> Also, have you run some benchmarks using the tilt environment and monitoring how the metrics look? I think it's worth doing so before merging.

>I only run benchmarks using the appender benchmarks in the repo but can you clarify this so we are aligned ? Are you asking to test the performance of the retry code or the overhead on the "normal operation" (requests succeed) ?

I do not want to speak for Marc, but IMO we should have some numbers on how this impacts overall performance. I would expect 
- no performance change if there are no retries
- retry behavior to add only a reasonable amount of overhead, but slowing down the processing of new events

Testing this e2e under load in a dev environment is essential before starting the promotion process.

**Comment by marclop** [comment]
@simitt, that's right.

@kruskall, I'll leave it up to you whether you want to test this as part of the dependency update or as part of the PR. If we go the dependency promotion path, we may approve the PR without any regression testing. I understand the retry behavior is opt-in rather than opt-out, but that's what regression testing is for.

**Review by marclop — APPROVED** [review]
LGTM. It'd be great to test this when either as part of the dependency update or before we merge.

File: bulk_indexer.go
Comment (axw) [inline bulk_indexer.go:323]: It looks like we're decompressing and scanning through the request body for every failed item. Should we be decompressing once, and scanning through incrementally? e.g. if you have failures at index 1 and 8, then first skip to the index 1, then skip from index 1 to index 8 without decompressing and reprocessing up to index 1.

**Review by axw — COMMENTED** [review]
If we don't need to export Indexnth then it would be better not to, to keep the public API tidy. The way the document indexes are found looks a bit inefficient, but we could improve it in a follow up since it should not be on a hot path.

**Review by inge4pres — APPROVED** [review]

File: bulk_indexer.go
Comment (axw) [reply bulk_indexer.go:323]: I don't mean decompress the whole payload at once, I meant create the decompressor once and stream through the results. AFAICS there's no need to call `gzip.NewReader` more than once.

Instead you could lazily create it when seeing the first 429 result, wrap it with [bufio.NewScanner](https://pkg.go.dev/bufio#NewScanner), and stream through the lines, skipping non-429 result lines as needed.

**Pull request merged**
